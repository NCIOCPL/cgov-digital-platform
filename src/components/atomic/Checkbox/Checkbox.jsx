import React from 'react';
import PropTypes from 'prop-types';
import './Checkbox.scss';

const Checkbox = ({
  id,
  label,
  value,
  name,
  classes,
  disabled,
  ...otherProps
}) => (
  <div className={`cts-checkbox ${classes}`}>
    <input
      id={id}
      className="cts-checkbox__input"
      type="checkbox"
      name={name}
      value={value ? value : id}
      disabled={disabled || false}
      {...otherProps}
    />
    <label className="cts-checkbox__label" htmlFor={id}>
      {label}
    </label>
  </div>
);

Checkbox.propTypes = {
  id: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string,
  value: PropTypes.string,
  disabled: PropTypes.bool,
  classes: PropTypes.string,
};

Checkbox.defaultProps = {
  classes: '',
  name: 'checkboxes',
};

export default Checkbox;
