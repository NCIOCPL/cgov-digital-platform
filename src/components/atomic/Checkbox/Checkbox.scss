.cts-checkbox {
  box-sizing: border-box;
  padding: 0;
  &__input {
    position: absolute;
    left: -999em;
  }
  &__label {
    cursor: pointer;
    display: inherit;
    font-weight: 500;
    margin-bottom: 0.75rem;
    padding-left: 2em;
    position: relative;
    text-indent: -2em;

    &::before {
      background: white;
      content: '\a0';
      display: inline-block;
      left: 2px;
      position: relative;
      vertical-align: middle;
      height: 1.25em;
      width: 1.25em;
      border-radius: 2px;
      box-shadow: 0 0 0 2px #71767a;
      line-height: 1.25rem;
      margin-right: 0.75rem;
    }
  }
}

.cts-checkbox__input:checked + .cts-checkbox__label::before {
  background-color: #005ea2;
  -webkit-box-shadow: 0 0 0 2px #005ea2;
  box-shadow: 0 0 0 2px #005ea2;
}

.cts-checkbox__input:checked + .cts-checkbox__label::before,
.cts-checkbox__input:checked:disabled + .cts-checkbox__label::before {
  background-image: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMCIgaGVpZ2h0PSIyMCIgdmlld0JveD0iMCAwIDIxNiAxNDYiPjxwYXRoIGZpbGw9IiNmZmYiIGQ9Ik0xNjguODYgMzcuOTY2bC0xMS4wOC0xMS4wOGMtMS41Mi0xLjUyLTMuMzY3LTIuMjgtNS41NC0yLjI4LTIuMTcyIDAtNC4wMi43Ni01LjU0IDIuMjhMOTMuMjU0IDgwLjQxNCA2OS4zIDU2LjM4Yy0xLjUyLTEuNTIyLTMuMzY3LTIuMjgyLTUuNTQtMi4yODItMi4xNzIgMC00LjAyLjc2LTUuNTQgMi4yOEw0Ny4xNCA2Ny40NmMtMS41MiAxLjUyMi0yLjI4IDMuMzctMi4yOCA1LjU0MiAwIDIuMTcyLjc2IDQuMDIgMi4yOCA1LjU0bDI5LjQ5MyAyOS40OTMgMTEuMDggMTEuMDhjMS41MiAxLjUyIDMuMzY4IDIuMjggNS41NCAyLjI4IDIuMTczIDAgNC4wMi0uNzYgNS41NC0yLjI4bDExLjA4Mi0xMS4wOEwxNjguODYgNDkuMDVjMS41Mi0xLjUyIDIuMjgzLTMuMzcgMi4yODMtNS41NCAwLTIuMTc0LS43Ni00LjAyLTIuMjgtNS41NHoiLz48L3N2Zz4='),
    linear-gradient(transparent, transparent);
  background-repeat: no-repeat;
  background-position: 50%;
}

.cts-checkbox__input:disabled + .cts-checkbox__label {
  color: #c9c9c9;
}

.cts-checkbox__input:focus + .cts-checkbox__label::before {
  outline: 0.25rem solid #2491ff;
  outline-offset: 0;
}

.cts-checkbox__input:disabled + .cts-checkbox__label::before {
  background: #e6e6e6;
  -webkit-box-shadow: 0 0 0 2px #c9c9c9;
  box-shadow: 0 0 0 2px #c9c9c9;
  cursor: not-allowed;
}

@media print {
  .cts-checkbox__input:checked + .cts-checkbox__label::before,
  .cts-checkbox__input:checked:disabled + .cts-checkbox__label::before {
    background-image: none;
    background-color: white;
    content: url('data:image/svg+xml;utf8,<svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 216 146"><path fill="#1b1b1b" d="M168.86 37.966l-11.08-11.08c-1.52-1.52-3.367-2.28-5.54-2.28-2.172 0-4.02.76-5.54 2.28L93.254 80.414 69.3 56.38c-1.52-1.522-3.367-2.282-5.54-2.282-2.172 0-4.02.76-5.54 2.28L47.14 67.46c-1.52 1.522-2.28 3.37-2.28 5.542 0 2.172.76 4.02 2.28 5.54l29.493 29.493 11.08 11.08c1.52 1.52 3.368 2.28 5.54 2.28 2.173 0 4.02-.76 5.54-2.28l11.082-11.08L168.86 49.05c1.52-1.52 2.283-3.37 2.283-5.54 0-2.174-.76-4.02-2.28-5.54z"/></svg>');
    text-indent: 0;
  }
}
