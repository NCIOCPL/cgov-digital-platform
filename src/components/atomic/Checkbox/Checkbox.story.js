import generateStories from '../../../../.storybook/generateStories';
import Checkbox from './Checkbox';

generateStories({
  component: { Checkbox },
  storyName: 'Elements|Checkbox',
  defaultProps: {
    id: 'storybook-checkbox',
    label: 'Example Checkbox',
    value: null,
    name: 'storybook-checkbox'
  },
  stories: [
    {
      name: 'default'
    }
  ]
});
