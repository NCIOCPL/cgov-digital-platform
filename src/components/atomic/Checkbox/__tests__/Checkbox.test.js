import React from 'react';
import { render, mount } from 'enzyme';
import Checkbox from '../Checkbox';

jest.unmock('../Checkbox.jsx');

const defaultProps = {
  label: 'label goes here',
  id: 'mock-checkbox',
};

const setup = (enzymeMethod = render, props = {}) => {
  const component = enzymeMethod(<Checkbox {...props} />);
  return component;
};

describe('Checkbox', () => {
  describe('Render', () => {

    it('renders without error', () => {
      const component = setup(render, defaultProps);
      expect(component).toMatchSnapshot();
    });
  });

  it('is defined', () => {
    const wrapper = setup(render, defaultProps);
    expect(wrapper).toBeDefined();
  });

  it('renders the label text correctly', () => {
    const wrapper = setup(mount, defaultProps);
    const labelElement = wrapper.find('label');
    expect(labelElement.text()).toBe(defaultProps.label);
  });

  it('has an id attribute', () => {
    const wrapper = setup(mount, defaultProps);
    expect(typeof wrapper.find('input').props().id).toBe('string');
  });

  it('has a matching id and htmlFor attributes', () => {
    const wrapper = setup(mount, defaultProps);
    const inputId = wrapper.find('input').props().id;
    const labelHtmlFor = wrapper.find('label').props().htmlFor;
    expect(labelHtmlFor).toBe(inputId);
  });

  it('has a default value property', () => {
    const wrapper = setup(mount, defaultProps);
    const id = wrapper.find('input').props().id;
    const value = wrapper.find('input').props().value;
    expect(value).toBe(id);
  });

  it('accepts a value property', () => {
    const testValue = 'testing';
    const wrapper = setup(mount, {
      ...defaultProps,
      value: testValue
    });
    const inputValue = wrapper.find('input').props().value;
    expect(inputValue).toBe(testValue);
  });

  it('has a default name property', () => {
    const wrapper = setup(mount, defaultProps);
    const inputName = wrapper.find('input').props().name;
    expect(inputName).toBe('checkboxes');
  });

  it('accepts a name property', () => {
    const testValue = 'testing';
    const wrapper = setup(mount, {
      ...defaultProps,
      name: testValue
    });
    const inputName = wrapper.find('input').props().name;
    expect(inputName).toBe(testValue);
  });

  it('it is not checked by default', () => {
    const wrapper = setup(mount, defaultProps);
    const inputChecked = wrapper.find('input').props().defaultChecked;
    expect(inputChecked).toBe(false);
  });

  it('accepts the defaultChecked property', () => {
    const wrapper = setup(mount, {
      ...defaultProps,
      defaultChecked: true
    });
    const inputChecked = wrapper.find('input').props().defaultChecked;
    expect(inputChecked).toBe(true);
  });

  it('is not disabled by default', () => {
    const wrapper = setup(mount, defaultProps);
    const inputDisabled = wrapper.find('input').props().disabled;
    expect(inputDisabled).toBe(false);
  });

  it('accepts the disabled property', () => {
    const wrapper = setup(mount, {
      ...defaultProps,
      disabled: true
    });
    const inputDisabled = wrapper.find('input').props().disabled;
    expect(inputDisabled).toBe(true);
  });
});
