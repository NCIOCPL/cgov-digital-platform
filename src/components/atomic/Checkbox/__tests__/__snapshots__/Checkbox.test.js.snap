// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Render renders without error 1`] = `
<div
  class="cts-checkbox "
>
  <input
    class="cts-checkbox__input"
    id="mock-checkbox"
    name="checkboxes"
    type="checkbox"
    value="mock-checkbox"
  />
  <label
    class="cts-checkbox__label"
    for="mock-checkbox"
  >
    label goes here
  </label>
</div>
`;
